src/ecosystem/corals/CoralGrowth.jl

```diff
@@ -25,33 +25,37 @@ function CoralGrowth(n_locs::Int64, n_groups::Int64, n_sizes::Int64)::CoralGrowt
-    # These are specific to the 35 "species"/ 5 group formulation
-    small::SVector = @SVector [1, 8, 15, 22, 29]
-    mid::SVector = SVector{25}(collect([2:6; 9:13; 16:20; 23:27; 30:34]))
-    large::SVector = @SVector [7, 14, 21, 28, 35]
-
-    # Values for ode_p
-    # TODO: The named tuple was for use with ODE solvers, which is no unneeded.
-    # These caches should all be moved into the `CoralGrowth` struct.
+    # The following splits into "small, mid and large" indices based on number of size classes and functional groups
+    indices_array = zeros(Int64, (n_groups, n_sizes))
+    for n_g in 0:(n_groups - 1)
+        indices_array[n_g + 1, :] .= collect((n_g * n_sizes + 1):(n_g * n_sizes + n_sizes))
+    end
+
+    mid_ind = collect(indices_array[:, 2:(end - 1)])[:]
+    size_mid = prod(size(indices_array[:, 2:(end - 1)]))
+
+    small::SVector{n_groups} = indices_array[:, 1]
+    mid::SVector{size_mid} = mid_ind
+    large::SVector{n_groups} = indices_array[:, end]
     p = @NamedTuple{
-        small::StaticArrays.SVector{5,Int64},  # indices for small size classes
-        mid::StaticArrays.SVector{25,Int64},   # indices for mid-size corals
-        large::StaticArrays.SVector{5,Int64},  # indices for large corals
-        rec::Matrix{Float64},                  # recruitment values, where `s` relates to available space (not max carrying capacity)
-        sXr::Array{Float64,3},                  # s * X * r
-        X_mb::Array{Float64,3},                 # X * mb
-        r::Matrix{Float64},                    # growth rate
-        mb::Matrix{Float64}                    # background mortality
+        small::StaticArrays.SVector{n_groups,Int64},           # indices for small size classes
+        mid::StaticArrays.SVector{size_mid,Int64},            # indices for mid-size corals
+        large::StaticArrays.SVector{n_groups,Int64},           # indices for large corals
+        rec::Matrix{Float64},                           # recruitment values, where `s` relates to available space (not max carrying capacity)
+        sXr::Array{Float64},                           # s * X * r
+        X_mb::Array{Float64},                          # X * mb
+        r::Matrix{Float64},                             # growth rate
+        mb::Matrix{Float64}                             # background mortality
     }((                        # cache matrix to hold X (current coral cover)
-        # Cached indices
+        # cached indices
         small, mid, large,
 
-        # Cache matrices
-        zeros(n_groups, n_locs),           # rec
-        zeros(n_groups, n_sizes, n_locs),  # sXr
+        # cache matrices
+        zeros(n_groups, n_locs),  # rec
+        zeros(n_groups, n_sizes, n_locs), # sXr
         zeros(n_groups, n_sizes, n_locs),  # X_mb
-        zeros(n_groups, n_sizes),          # r
-        zeros(n_groups, n_sizes)           # mb
+        zeros(n_groups, n_sizes),  # r
+        zeros(n_groups, n_sizes)   # mb
```

Comment:
This entire bit needs to be revisited as many of these cache are not used any more (they were for the older ODE method).